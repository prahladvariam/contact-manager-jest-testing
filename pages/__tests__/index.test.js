import renderer from "react-test-renderer";
import ContactPage, {getStaticProps} from "../contacts";
import React from 'react';
import { act } from "react-dom/test-utils";
import { render, fireEvent, screen, waitFor } from '@testing-library/react';

describe("Testing .....", () => {
    const contactsList = [
        {
            id: 1,
            name: "Sample1",
            email: "sample1@mail.com",
            mobile: 5451651845
        },
        {
            id: 2,
            name: "Sample2",
            email: "sample2@mail.com",
            mobile: 5451651844555
        }
    ]

    beforeEach(() => {
        jest.resetModules();
    })
    
    const unmockedFetch = global.fetch

    beforeAll(() => {
        jest.setTimeout(90 * 1000)
        global.fetch = () =>
        Promise.resolve({
        json: () => Promise.resolve([]),
        })
    })

    afterAll(() => {
        global.fetch = unmockedFetch
    })

    it("component has Email option",async() => {
        
        const component = renderer.create(<ContactPage contactsList={contactsList} />).root
        expect(component.findByProps({className: "Email"}).children).toEqual(['Email'])
        await act(() => Promise.resolve())
    });
    
    it("component rendering",async() => {
        
        const component = render(<ContactPage contactsList={[]} />)
        expect(component).toMatchSnapshot()
        await act(() => Promise.resolve())
    });
    
    test("setContact is called and sorting is working",async() => {
        const mocksetState = jest.fn(() => Promise.resolve);
        const useStateSpy = jest.spyOn(React, "useState");
        useStateSpy.mockImplementation((initialState) => [initialState, mocksetState]);
        render(<ContactPage contactsList={[]} />)
        const Email = screen.getByTestId('email-test')
        fireEvent.click(Email)
        expect(mocksetState).toHaveBeenCalled();
        await act(() => Promise.resolve())
        
    });
    
    test("Search is working...",async() => {
        const mocksetState = jest.fn(() => Promise.resolve);
        const useStateSpy = jest.spyOn(React, "useState");
        useStateSpy.mockImplementation((initialState) => [initialState, mocksetState]);
        const component = renderer.create(<ContactPage contactsList={contactsList} />).root
        component.findByProps({className: "form-control m-2"}).onChange
        expect(mocksetState).toHaveBeenCalled();
        await act(() => Promise.resolve())
        
    });

    test("Delete is working...",async() => {
        const mocksetState = jest.fn(() => Promise.resolve);
        const useStateSpy = jest.spyOn(React, "useState");
        useStateSpy.mockImplementation((initialState) => [initialState, mocksetState]);
        const component = renderer.create(<ContactPage contactsList={contactsList} />).root
        component.findByProps({id: "Deletebtn1"}).onClick
        expect(mocksetState).toHaveBeenCalled();
        await act(() => Promise.resolve())
        
    });

    test("getStaticProps working..", async () => {
        const response =  await getStaticProps()
        expect(response).not.toBe(null)
        await act(() => Promise.resolve())
    })

    
})
