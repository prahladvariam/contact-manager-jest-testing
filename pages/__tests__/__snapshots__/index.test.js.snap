// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ..... component rendering 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container-fluid"
      >
        <div
          class="row"
        >
          <h2
            class="text-center bg-dark text-white"
          >
            CONTACT MANAGER
          </h2>
        </div>
        <div
          class="container"
        >
          <div
            class="row"
          >
            <div
              style="float: left; width: 50%;"
            >
              <h2>
                Contact List
              </h2>
            </div>
            <div
              style="float: right; width: 50%;"
            >
              <button
                class="btn btn-primary m-2 float-end"
                id="AddContactLink"
              >
                Add Contact
              </button>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              style="float: left; width: 20%;"
            >
              <input
                class="form-control m-2"
                placeholder="Search Contacts"
                type="text"
                value=""
              />
            </div>
            <div
              style="float: none; width: 20%;"
            >
              <button
                class="btn btn-primary m-2"
              >
                Search
              </button>
            </div>
            <div
              style="float: right; width: 60%;"
            >
              <h2
                class="float-end"
              >
                <select
                  class="form-select m-2 float-end"
                >
                  <option>
                    Sort
                  </option>
                  <option
                    value="name"
                  >
                    Name
                  </option>
                  <option
                    class="Email"
                    data-testid="email-test"
                    value="email"
                  >
                    Email
                  </option>
                </select>
              </h2>
            </div>
          </div>
          <div
            class="row"
          >
            <div />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container-fluid"
    >
      <div
        class="row"
      >
        <h2
          class="text-center bg-dark text-white"
        >
          CONTACT MANAGER
        </h2>
      </div>
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            style="float: left; width: 50%;"
          >
            <h2>
              Contact List
            </h2>
          </div>
          <div
            style="float: right; width: 50%;"
          >
            <button
              class="btn btn-primary m-2 float-end"
              id="AddContactLink"
            >
              Add Contact
            </button>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            style="float: left; width: 20%;"
          >
            <input
              class="form-control m-2"
              placeholder="Search Contacts"
              type="text"
              value=""
            />
          </div>
          <div
            style="float: none; width: 20%;"
          >
            <button
              class="btn btn-primary m-2"
            >
              Search
            </button>
          </div>
          <div
            style="float: right; width: 60%;"
          >
            <h2
              class="float-end"
            >
              <select
                class="form-select m-2 float-end"
              >
                <option>
                  Sort
                </option>
                <option
                  value="name"
                >
                  Name
                </option>
                <option
                  class="Email"
                  data-testid="email-test"
                  value="email"
                >
                  Email
                </option>
              </select>
            </h2>
          </div>
        </div>
        <div
          class="row"
        >
          <div />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
